### PR TITLE
feat(app): Add Labware Detail with Well Row/Column Highlighting

### LINKED ISSUE #8380: Labware Position Check: Labware Detail Well Row/Column Highlighting
As shown in this figma [node](https://www.figma.com/file/JCQNOTTK9tTmYPhRqeOfSj/Milestone-0:-Protocol-Upload-Revamp?node-id=4327%3A837315), implement dynamic highlighting of the well labels and well outlines of the labware detailed within the generic step component.

### Acceptance Criteria

- [ ] well(s) accessed by the nozzle(s) of the primary pipette should be highlighted with stroke color and well label text color `C_BLUE`.

### PR DESCRIPTION
**Pull request by shlokamin**
# Overview

This PR adds well row/column highlighting, and labware outline coloring for the LPC flow. 

closes #8380

# Changelog
- Adds Labware Detail with Well Row/Column Highlighting

# Review requests

- code review
- make sure labware + pipette renders show up in LPC flow

# Risk assessment
Low

### CODE REVIEW

File: app/src/assets/localization/en/labware_position_check.json [hunk 1]
```diff
@@ -8,10 +8,10 @@
...
-  "labware_step_detail_labware_plural": "The tips should be centered above column 1 in {{labware_name}} and <bold> level </bold> with the top of the labware.",
-  "labware_step_detail_tiprack": "The pipette nozzle should be centered above A1 in {{tiprack_name}} and <bold> level </bold> with the top of the tip.",
-  "labware_step_detail_tiprack_plural": "The pipette nozzles should be centered above column 1 in {{tiprack_name}} and <bold> level </bold> with the top of the tips.",
+  "labware_step_detail_labware": "The tip should be centered above A1 in {{labware_name}} and <bold>level</bold> with the top of the labware.",
+  "labware_step_detail_labware_plural": "The tips should be centered above column 1 in {{labware_name}} and <bold>level</bold> with the top of the labware.",
+  "labware_step_detail_tiprack": "The pipette nozzle should be centered above A1 in {{tiprack_name}} and <bold>level</bold> with the top of the tip.",
+  "labware_step_detail_tiprack_plural": "The pipette nozzles should be centered above column 1 in {{tiprack_name}} and <bold>level</bold> with the top of the tips.",
```
Comment (shlokamin) [inline app/src/assets/localization/en/labware_position_check.json:14]: removed the unnecessary whitespace inside of the `bold` elements, discovered they were unnecessary by trying to verify text in a unit test

File: app/src/organisms/ProtocolSetup/LabwarePositionCheck/LabwarePositionCheckStepDetail.tsx [hunk 2]
```diff
@@ -6,28 +6,94 @@ import {
...
+      marginLeft="30%"
       boxShadow="1px 1px 1px rgba(0, 0, 0, 0.25)"
       borderRadius="4px"
       backgroundColor={C_NEAR_WHITE}
+      flexDirection={'column'}
     >
-      <StepDetailText selectedStep={props.selectedStep} />
+      <StepDetailText
+        selectedStep={props.selectedStep}
+        pipetteChannels={pipetteChannels}
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/LabwarePositionCheck/LabwarePositionCheckStepDetail.tsx:79]: feeding in the number of pipette channels explicitly here. inferring this from `useIntroInfo` doesn't actually work because `numberOfTips` from `useIntroInfo` just gets the number of tips from the primary pipette. 

`StepDetailText` needs to be dynamic depending on which specific section we're in

File: app/src/organisms/ProtocolSetup/LabwarePositionCheck/StepDetailText.tsx [hunk 3]
```diff
@@ -8,31 +8,31 @@ import {
...
   const [
     showLabwarePositionCheckStepDetailModal,
     setLabwarePositionCheckStepDetailModal,
   ] = React.useState<boolean>(false)
-  if (protocolData == null || introInfo == null) return null
+  if (protocolData == null) return null
   const labwareDefId = protocolData.labware[labwareId].definitionId
   const displayName =
     protocolData.labwareDefinitions[labwareDefId].metadata.displayName
-  const { numberOfTips } = introInfo
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/LabwarePositionCheck/StepDetailText.tsx:35]: this doesn't work, see comment above

File: app/src/organisms/ProtocolSetup/LabwarePositionCheck/__tests__/StepDetailText.test.tsx [hunk 4]
```diff
@@ -131,63 +113,63 @@ describe('StepDetailText', () => {
...
+    }
     const { getByText } = render(props)
     getByText('See how to tell if the pipette is centered')
-    nestedTextMatcher(
-      'The tip should be centered above A1 in NEST 96 Well Plate 100 µL PCR Full Skirt and level with the top of the labware.'
+    getByText(
+      nestedTextMatcher(
+        'The tip should be centered above A1 in NEST 96 Well Plate 100 µL PCR Full Skirt and level with the top of the labware.'
+      )
     )
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/LabwarePositionCheck/__tests__/StepDetailText.test.tsx:136]: these tests were always passing before. `nestedTextMatcher ` has to be used in conjunction with a RTL query

File: components/src/hardware-sim/Labware/labwareInternals/LabwareOutline.tsx [hunk 5]
```diff
@@ -3,13 +3,14 @@ import cx from 'classnames'
...
 
+import type { CSSProperties } from 'styled-components'
 import type { LabwareDefinition2 } from '@opentrons/shared-data'
-
 export interface LabwareOutlineProps {
   definition?: LabwareDefinition2
   width?: number
   height?: number
   isTiprack?: boolean
+  stroke?: CSSProperties['stroke']
```
Comment (shlokamin) [inline components/src/hardware-sim/Labware/labwareInternals/LabwareOutline.tsx:13]: added this because the labware outline color has to be customizable for LPC. but this prop will also be handy in the app renav project

File: components/src/hardware-sim/Labware/labwareInternals/StaticLabware.tsx [hunk 6]
```diff
@@ -34,24 +33,13 @@ const TipDecoration = React.memo(function TipDecoration(props: {
   return null
 })
 
-function StaticLabwareComponent(props: StaticLabwareProps): JSX.Element {
+export function StaticLabwareComponent(props: StaticLabwareProps): JSX.Element {
```
Comment (shlokamin) [inline components/src/hardware-sim/Labware/labwareInternals/StaticLabware.tsx:36]: exporting so i can use this non memoized component in the test. for some reason jest doesn't like it when I mock a memoized react component 🤔

File: components/src/hardware-sim/Labware/labwareInternals/StaticLabware.tsx [hunk 7]
```diff
@@ -34,24 +33,13 @@ const TipDecoration = React.memo(function TipDecoration(props: {
...
-          x={1}
-          y={1}
-          strokeWidth={1}
-          width={props.definition.dimensions.xDimension - 2}
-          height={props.definition.dimensions.yDimension - 2}
-          rx={6}
-          className={cx(styles.labware_outline, {
-            [styles.tiprack_outline]: isTiprack,
-          })}
-        />
```
Comment (shlokamin) [inline components/src/hardware-sim/Labware/labwareInternals/StaticLabware.tsx:54]: so I'm actually not sure why this was here to begin with... it is redundant with what `LabwareOutline` does. It seems like this may have been left by accident?

File: components/src/hardware-sim/Labware/labwareInternals/WellLabels.tsx [hunk 8]
```diff
@@ -1,56 +1,111 @@
...
+  const fillColor = wellLabelColor ?? C_BLACK
+  const LETTER_COLUMN_X =
+    wellLabelOption === WELL_LABEL_OPTIONS.SHOW_LABEL_INSIDE
+      ? LETTER_COLUMN_X_INSIDE
+      : LETTER_COLUMN_X_OUTSIDE
+
+  const NUMBER_COLUMN_Y_FROM_TOP =
+    wellLabelOption === WELL_LABEL_OPTIONS.SHOW_LABEL_INSIDE
+      ? NUMBER_COLUMN_Y_FROM_TOP_INSIDE
+      : NUMBER_COLUMN_Y_FROM_TOP_OUTSIDE
```
Comment (shlokamin) [inline components/src/hardware-sim/Labware/labwareInternals/WellLabels.tsx:43]: this is so we can dynamically place well labels depending on whether the consumer wants them inside or outside the labware outline

File: components/src/testing/utils/matchers.ts [hunk 9]
```diff
@@ -13,3 +14,23 @@ export const partialComponentPropsMatcher = (argsToMatch: unknown): any =>
...
+    return typeof textMatch === 'string'
+      ? Boolean(n?.textContent.match(textMatch))
+      : textMatch.test(n.textContent)
+  }
+  const nodeHasText = hasText(node)
+  const childrenDontHaveText =
+    node != null && Array.from(node.children).every(child => !hasText(child))
+
+  return nodeHasText && childrenDontHaveText
+}
```
Comment (shlokamin) [inline components/src/testing/utils/matchers.ts:36]: moved this here to be with the other test matchers

File: app/src/organisms/ProtocolSetup/LabwarePositionCheck/__tests__/StepDetailText.test.tsx (hunk 4 above)
Comment (jerader) [reply app/src/organisms/ProtocolSetup/LabwarePositionCheck/__tests__/StepDetailText.test.tsx:136]: thanks for fixing this!

File: components/src/hardware-sim/Labware/LabwareRender.tsx [hunk 10]
```diff
@@ -3,25 +3,47 @@ import {
...
   selectedWells?: WellGroup | null | undefined
   /** CSS color to fill specified wells */
   wellFill?: WellFill
+  /** CSS color to stroke specified wells */
+  wellStroke?: WellStroke
+  /** CSS color to stroke the labware outline */
+  labwareStroke?: CSSProperties['stroke']
   /** Optional callback, called with WellMouseEvent args onMouseEnter */
   onMouseEnterWell?: (e: WellMouseEvent) => unknown
   /** Optional callback, called with WellMouseEvent args onMouseLeave */
```
Comment (jerader) [inline components/src/hardware-sim/Labware/LabwareRender.tsx:49]: thanks for leaving all these comments in here. makes it a lot easier to understand what is happening 😄

File: components/src/hardware-sim/Labware/labwareInternals/WellLabels.tsx (hunk 8 above)
Comment (jerader) [reply components/src/hardware-sim/Labware/labwareInternals/WellLabels.tsx:43]: this is cool, i didn't realize this is an option that we wanted.

File: app/src/organisms/ProtocolSetup/LabwarePositionCheck/StepDetailText.tsx [hunk 11]
```diff
@@ -41,15 +41,19 @@ export const StepDetailText = (
...
           i18nKey={
             displayName.includes('Tip Rack')
               ? 'labware_step_detail_tiprack'
               : 'labware_step_detail_labware'
           }
-          count={numberOfTips}
+          count={pipetteChannels}
           values={{ labware_name: displayName, tiprack_name: displayName }}
           components={{
             bold: <strong />,
```
Comment (jerader) [inline app/src/organisms/ProtocolSetup/LabwarePositionCheck/StepDetailText.tsx:58]: the link size doesn't match up with the text size.
Screen Shot 2021-10-11 at 08 02 55

**Review by jerader — APPROVED** [review]
Looking good! Great job! I left a sprinkle of comments. Super excited to see the `labwarePositionCheckStepDetail` component come together! Also, thanks for fixing up the nested text matcher and for putting it into the `matchers.ts` file

One thing I thought I'd mention: since we removed the usage of the `useIntroInfo` hook in `StepDetailText` and since `IntroScreen.tsx` no longer uses that hook since the text is now static, that hook is currently not in use. If it ends up being something we don't use at all in the end with LPC, is it good practice to remove it? or just keep it in case?

File: components/src/hardware-sim/Labware/labwareInternals/WellLabels.tsx (hunk 8 above)
Comment (shlokamin) [reply components/src/hardware-sim/Labware/labwareInternals/WellLabels.tsx:43]: yeah! we havent ever had to render well labels outside of labware before, this is the first time

**Comment by shlokamin** [comment]
> One thing I thought I'd mention: since we removed the usage of the useIntroInfo hook in StepDetailText and since IntroScreen.tsx no longer uses that hook since the text is now static, that hook is currently not in use.

@jerader isn't `IntroScreen.tsx` still using it to grab this stuff? 

```js
const {
    primaryPipetteMount,
    secondaryPipetteMount,
    firstStepLabwareSlot,
    sections,
  } = introInfo
```

i think the hook is now doing more stuff than it needs to so we can delete some of the logic, but i think we still need the info above
